engines/docker/image_cache.go

```diff
@@ -0,0 +1,103 @@
+package dockerengine
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+	"github.com/taskcluster/taskcluster-worker/runtime/gc"
+)
+
+const dockerPullImageInactivityTimeout = 30 * time.Second
+
+type imageResource struct {
+	Repository string
+	Tag        string
+	Size       uint64
+	dispose    atomics.Once
+}
+
+type cacheContext struct {
+	context.Context
+}
+
+func newCacheContext(ctx context.Context) *cacheContext {
+	if ctx != nil {
+		return &cacheContext{
+			Context: ctx,
+		}
+	}
+	return nil
+}
+
+func (ctx *cacheContext) Progress(description string, percent float64) {
+	debug("progress: %s %f", description, percent)
+}
+
+func (im *imageResource) MemorySize() (uint64, error) {
+	return im.Size, nil
+}
+
+func (im *imageResource) DiskSize() (uint64, error) {
+	return 0, gc.ErrDisposableSizeNotSupported
+}
+
+func (im *imageResource) Dispose() error {
+	im.dispose.Do(func() {
+		client, err := docker.NewClientFromEnv()
+		if err != nil {
+			debug("%v", err)
+			return
+		}
+		err = client.RemoveImage(buildImageName(im.Repository, im.Tag))
+		if err != nil {
+			debug("%v", err)
+		}
+	})
+	return nil
+}
+
+func imageConstructor(ctx caching.Context, opts interface{}) (caching.Resource, error) {
+	options := opts.(imageType)
+	client, err := docker.NewClientFromEnv()
+	if err != nil {
+		return nil, err
+	}
+
+	debug("pulling image %s %s", options.Repository, options.Tag)
+	err = client.PullImage(docker.PullImageOptions{
+		Repository:        options.Repository,
+		Tag:               options.Tag,
+		InactivityTimeout: dockerPullImageInactivityTimeout,
+		Context:           ctx,
```

Comment:
Add a TODO to use `OutputStream` to read progress and report it to `ctx`.
(this is a nice to have, we can develop a helper function that can do this for us later, in another PR).

Reply:
Good point...ctx is a *caching.Context here so this may not be too hard to do in the future